# mehmetfatih

In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
from sklearn import linear_model
import warnings
warnings.filterwarnings('ignore')

In [2]:
prices = pd.read_csv('dataset/house_prices.csv', index_col=0)
prices.tail(2)

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
1459          20       RL         68.0     9717   Pave   NaN      Reg   
1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolArea PoolQC Fence MiscFeature  \
Id                                    ...                                     
1459         Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
1460         Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   

     MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                               
1459       0      4    2010        WD         Normal     142125  
1460       0      6    2008        WD         Normal     147500  

[2 rows x 80 columns]

## Q1 :
Ev fiyatları modelinizi tekrar çalıştırın ve F-testi, R-kare, ayarlanmış R-kare, AIC ve BIC kullanarak modelinizin uygunluğunu değerlendirin.

In [3]:
high=prices.corr().SalePrice.abs().sort_values(ascending=False).index[0:11]

In [4]:
Y = prices['SalePrice']
X = prices[['OverallQual', 'GrLivArea', 'GarageCars', 'GarageArea',
       'TotalBsmtSF', '1stFlrSF', 'FullBath', 'TotRmsAbvGrd', 'YearBuilt',
       'YearRemodAdd']]
lrm = linear_model.LinearRegression()
lrm.fit(X, Y)
print('Değişkenler: \n', lrm.coef_)
print('Sabit değer (bias): \n', lrm.intercept_)

Değişkenler: 
 [ 1.96045898e+04  5.12971178e+01  1.04179010e+04  1.49475334e+01
  1.98650991e+01  1.41737355e+01 -6.79087146e+03  3.31050771e+01
  2.68240707e+02  2.96481161e+02]
Sabit değer (bias): 
 -1186193.7666661367


In [5]:
X = sm.add_constant(X)

sonuclar = sm.OLS(Y, X).fit()
sonuclar.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.774
Model:                            OLS   Adj. R-squared:                  0.772
Method:                 Least Squares   F-statistic:                     495.4
Date:                Fri, 24 Jan 2020   Prob (F-statistic):               0.00
Time:                        17:27:16   Log-Likelihood:                -17459.
No. Observations:                1460   AIC:                         3.494e+04
Df Residuals:                    1449   BIC:                         3.500e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -1.186e+06   1.29e+05     -9.187      0.000   -1.44e+06   -9.33e+05
OverallQual    1.96e+04   1190.159     16.472      0.000    1.73e+04    2.19e+04
GrLivArea       51.2971      4.233     12.119      0.000      42.994      59.600
GarageCars    1.042e+04   3044.272      3.422      0.001    4446.249    1.64e+04
GarageArea      14.9475     10.311      1.450      0.147      -5.279      35.174
TotalBsmtSF     19.8651      4.295      4.625      0.000      11.439      28.291
1stFlrSF        14.1737      4.930      2.875      0.004       4.504      23.844
FullBath     -6790.8715   2682.369     -2.532      0.011   -1.21e+04   -1529.130
TotRmsAbvGrd    33.1051   1119.061      0.030      0.976   -2162.048    2228.258
YearBuilt      268.2407     50.346      5.328      0.000     169.481     367.000
YearRemodAdd   296.4812     63.635      4.659      0.000     171.655     421.307
==============================================================================
Omnibus:                      477.814   Durbin-Watson:                   1.985
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            58906.279
Skew:                          -0.412   Prob(JB):                         0.00
Kurtosis:                      34.107   Cond. No.                     4.69e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.69e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

""""R-squared:	0.774,
Adj. R-squared:	0.772"""
Bu, modelimizin fiyatlardaki varyansın %77,4'lük bir oranla açıkladığı, %22,6'sını ise açıklamadan bıraktığı anlamına gelir. Modelin yüksek oranla fiyatlardaki varyansı açıkladığı sonucuna varırız. Bununla birlikte modelimizin geliştirilmesine devam edilebilir.

## Q2 :
Modelinizin tatmin edici olduğunu düşünüyor musunuz? Açıklayın

""""R-squared:	0.774,
Adj. R-squared:	0.772"""
Bu, modelimizin fiyatlardaki varyansın %77,4'lük bir oranla açıkladığı, %22,6'sını ise açıklamadan bıraktığı anlamına gelir. Modelin yüksek oranla fiyatlardaki varyansı açıkladığı sonucuna varırız. Bununla birlikte modelimizin geliştirilmesine devam edilebilir. Modelimiz tahmin gücü yüksektir. Eğer R_squared fazla yükselmesi durumunda yanlılık ortaya çıkacaktır. Performansı yeterli buluyorum. Hiçbir örneklem veri kümesi gerçek popülasyon kadar iyi olamayacağını göz önünde bulundurmamız gerekmektedir. 
Ayrıca p değeri uygun olmayan değişkenler modelden çıkartılabilir. Zaten etkisiz olan iki değişkenin çıkması tahmin performansını değiştirmyecektir. 

## Q3:
Modelinizin uygunluğunu iyileştirmek için, bazı değişkenleri ekleyerek veya kaldırarak farklı model özelliklerini deneyin.

In [6]:
Y = prices['SalePrice']
X = prices[['OverallQual', 'GrLivArea', 'GarageCars',
       'TotalBsmtSF', '1stFlrSF', 'FullBath', 'YearBuilt',
       'YearRemodAdd']]
lrm = linear_model.LinearRegression()
lrm.fit(X, Y)
print('Değişkenler: \n', lrm.coef_)
print('Sabit değer (bias): \n', lrm.intercept_)

Değişkenler: 
 [ 1.95867751e+04  5.17669244e+01  1.39510419e+04  2.03846645e+01
  1.46511561e+01 -7.18377849e+03  2.69889971e+02  2.95740619e+02]
Sabit değer (bias): 
 -1188025.7254260296


In [7]:
X = sm.add_constant(X)

sonuclar = sm.OLS(Y, X).fit()
sonuclar.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.773
Model:                            OLS   Adj. R-squared:                  0.772
Method:                 Least Squares   F-statistic:                     618.9
Date:                Fri, 24 Jan 2020   Prob (F-statistic):               0.00
Time:                        17:27:16   Log-Likelihood:                -17460.
No. Observations:                1460   AIC:                         3.494e+04
Df Residuals:                    1451   BIC:                         3.499e+04
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -1.188e+06   1.28e+05     -9.255      0.000   -1.44e+06   -9.36e+05
OverallQual   1.959e+04   1188.062     16.486      0.000    1.73e+04    2.19e+04
GrLivArea       51.7669      3.097     16.714      0.000      45.692      57.842
GarageCars    1.395e+04   1816.623      7.680      0.000    1.04e+04    1.75e+04
TotalBsmtSF     20.3847      4.269      4.775      0.000      12.010      28.759
1stFlrSF        14.6512      4.919      2.979      0.003       5.002      24.300
FullBath     -7183.7785   2643.957     -2.717      0.007   -1.24e+04   -1997.392
YearBuilt      269.8900     50.169      5.380      0.000     171.478     368.302
YearRemodAdd   295.7406     63.618      4.649      0.000     170.947     420.534
==============================================================================
Omnibus:                      462.446   Durbin-Watson:                   1.984
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            54365.371
Skew:                          -0.355   Prob(JB):                         0.00
Kurtosis:                      32.886   Cond. No.                     4.62e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.62e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Ayrıca p değeri uygun olmayan değişkenler modelden çıkartılabilir. Zaten etkisiz olan iki değişkenin çıkması tahmin performansını değiştirmeyecektir. Ancak BIC değerindeki düşüşe dikkat etmek gerekir. Bu nedenle istatistiksel olarak anlamsız olan değişkenler modelden çıkarılmadılır. 

In [8]:
Y = prices['SalePrice']
prices['Garage']=prices['GarageCars']*prices['GarageArea']
X = prices[['OverallQual', 'GrLivArea', 'Garage',
       'TotalBsmtSF', '1stFlrSF', 'FullBath', 'YearBuilt',
       'YearRemodAdd']]
lrm = linear_model.LinearRegression()
lrm.fit(X, Y)
print('Değişkenler: \n', lrm.coef_)
print('Sabit değer (bias): \n', lrm.intercept_)

Değişkenler: 
 [ 1.92562697e+04  4.94421438e+01  1.96374373e+01  1.74900859e+01
  1.32099408e+01 -6.16031460e+03  2.69732384e+02  2.76898089e+02]
Sabit değer (bias): 
 -1136168.6429811432


In [9]:
X = sm.add_constant(X)

sonuclar = sm.OLS(Y, X).fit()
sonuclar.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.781
Model:                            OLS   Adj. R-squared:                  0.779
Method:                 Least Squares   F-statistic:                     645.3
Date:                Fri, 24 Jan 2020   Prob (F-statistic):               0.00
Time:                        17:27:16   Log-Likelihood:                -17437.
No. Observations:                1460   AIC:                         3.489e+04
Df Residuals:                    1451   BIC:                         3.494e+04
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -1.136e+06   1.26e+05     -9.029      0.000   -1.38e+06   -8.89e+05
OverallQual   1.926e+04   1163.663     16.548      0.000     1.7e+04    2.15e+04
GrLivArea       49.4421      3.066     16.127      0.000      43.428      55.456
Garage          19.6374      1.884     10.425      0.000      15.942      23.333
TotalBsmtSF     17.4901      4.204      4.161      0.000       9.244      25.736
1stFlrSF        13.2099      4.840      2.729      0.006       3.715      22.705
FullBath     -6160.3146   2600.683     -2.369      0.018   -1.13e+04   -1058.814
YearBuilt      269.7324     48.505      5.561      0.000     174.584     364.881
YearRemodAdd   276.8981     62.637      4.421      0.000     154.029     399.767
==============================================================================
Omnibus:                      501.219   Durbin-Watson:                   1.969
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            62062.676
Skew:                          -0.515   Prob(JB):                         0.00
Kurtosis:                      34.924   Cond. No.                     4.80e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.8e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

## Q4 :
Denediğiniz her model için, uygun metrikleri alın ve modellerinizi birbiriyle karşılaştırın. Hangi model en iyisidir ve neden?

Elimizdeki ilk modelde kullandığımız bir değişkeni ('GarageCar') başka bir değişkenle etkileşeme sokarak modelime ekledim. Bu sayede modelimde iyileşme gözlemlendi. Ayrıca yine istatiksel anlamı olmayan bir değişkeni ise modelimden tamamen çıkardım. Bu modelimin tahmin performansını yeterli bulmaktayım. İlk iki modele nispeten daha mükemmel sonuç elde ettim. 

WORK FİNİSHED.